### PR TITLE
Support PEFT with Liger in DPO

### PR DESCRIPTION
**Pull request by albertvillanova**
Support PEFT models with use_liger_kernel=True in `DPOTrainer`.

This PR enables PEFT models (e.g. LoRA) to be used with `use_liger_kernel=True` in `DPOTrainer`, lifting the blanket NotImplementedError that previously blocked all PEFT+Liger combinations.

### Motivation

The Liger fused DPO loss bypasses the model's `forward()` and multiplies hidden states by `lm_head.weight` directly. The previous guard raised `NotImplementedError` for any PEFT model, but this was too broad: the only genuinely incompatible case is when lm_head itself is wrapped by a PEFT adapter (e.g. "lm_head" in target_modules), because then lm_head.weight is the frozen base weight and the adapter delta is silently ignored. When lm_head is not adapted, PEFT+Liger works correctly.

### Compatibility and Error Handling

* Added a check to prevent using `use_liger_kernel=True` when a PEFT adapter is applied to `lm_head`, raising a clear error if this unsupported configuration is detected. This avoids silent failures where the head adapter would not be trained.
* Imported `BaseTunerLayer` from `peft.tuners.tuners_utils` to enable the above compatibility check.

### Model Unwrapping and Reference Handling

* Add PEFT double-unwrap (model = model.base_model.model) in _compute_loss_liger before backbone resolution, mirroring GRPOTrainer._get_last_hidden_state
* Handle self.ref_model is None in _compute_loss_liger (PEFT with no explicit reference model) by recovering reference behaviour via adapter disabling/switching, consistent with the existing _compute_loss logit path

---

> [!NOTE]
> **Medium Risk**
> Changes core DPO loss/reference computation for PEFT+Liger; incorrect unwrapping or ref-adapter handling could skew training, though narrow compatibility checks limit exposure.
> 
> **Overview**
> **`DPOTrainer` no longer rejects every PEFT model when `use_liger_kernel=True`.** LoRA (and similar weight adapters) that leave `lm_head` as a plain linear layer can train with the fused Liger DPO path.
> 
> Init-time guards replace the old blanket error: **`lm_head` in PEFT `target_modules`** raises because Liger reads frozen `lm_head.weight` and would skip head adapters; **prompt-learning PEFT** (e.g. Prompt Tuning) raises because Liger calls the backbone directly and never prepends virtual tokens.
> 
> In **`_compute_loss_liger`**, PEFT models are unwrapped to the inner backbone before the forward pass, and when there is **no separate `ref_model`**, reference hidden states are computed via **`use_adapter`** (ref adapter or adapters disabled), matching the standard DPO reference path.
> 
> Tests add an end-to-end **Liger + LoRA** training check and two **fail-fast** cases for `lm_head` LoRA and prompt tuning.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a624c49ab0e1be4ec34655603c512bb198764579. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -753,7 +754,19 @@ def __init__(
...
+                # head adapter would silently receive no gradient, so the model trains as if `lm_head` were frozen.
+                # Fail loudly rather than train a silently-frozen head.
+                output_embeddings = model.get_output_embeddings()
+                if isinstance(output_embeddings, BaseTunerLayer):
+                    raise ValueError(
+                        "`use_liger_kernel=True` is incompatible with applying a PEFT adapter to `lm_head`. The Liger "
+                        "fused DPO loss reads `lm_head.weight` directly, so the adapter on the head is ignored and "
+                        "never trained. Either remove `'lm_head'` from your `target_modules`, or set "
+                        "`use_liger_kernel=False`."
+                    )
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py:772]: OK, I'm implementing the guard differently. Additionally, I'm checking if other trainers need this guard as well.

File: tests/test_dpo_trainer.py [hunk 2]
```diff
@@ -787,25 +788,6 @@ def test_train_with_liger(self):
...
-
-        with pytest.raises(NotImplementedError, match="Liger DPO loss is not implemented for PEFT models."):
-            DPOTrainer(
-                model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
-                args=training_args,
-                train_dataset=dataset,
-                peft_config=LoraConfig(),
-            )
-
     def test_train_with_iterable_dataset(self):
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply tests/test_dpo_trainer.py:870]: I'm adding tests.

File: trl/trainer/dpo_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py:772]: I opened:
- https://github.com/huggingface/trl/pull/6186

**Review by qgallouedec — APPROVED** [review]
It looks good!
